### ISSUE TITLE
False positive 'This condition will always return false' comparing `this` with an instance of a subclass

### ISSUE THREAD

**Issue by yujiang** [microsoft/TypeScript#46709]
# Bug Report

### 💻 Code
```
class AA {
	do1 (){
		const b = dd.getB();
		if (this === b){
			console.log("this === b");
		}
	}
}

class BB extends AA{
	getB():BB {return this;}
}

let dd = new BB();
dd.do1();
```

### 🙁 Actual behavior
at line: 		if (this === b){
error TS2367: This condition will always return 'false' since the types 'this' and 'BB' have no overlap.

and execute the compiled js file will output:
this === b

**Comment by connordavison** [comment]
More generally, TS2367 has unexpected behaviour.

```ts
type A = { foo(): void; };
type B = { bar(): void; };
declare const a: A;
declare const b: B;
a === b; // TS2367 This comparison appears to be unintentional because the types 'A' and 'B' have no overlap.
```

In the above, `A` and `B` are compatible, and so `a === b` is possible. It's hard to tell otherwise if it's correct. This case can reasonably occur in a mixin-based system.

I _would_ expect an error if `A` and `B` were conflicting types such that `a === b` were impossible, e.g.:

```ts
type A = { foo(x: number): string; };
type B = { foo(y: string): number };
declare const a: A;
declare const b: B;
a === b; // TS2367 This comparison appears to be unintentional because the types 'A' and 'B' have no overlap.
```

**Comment by RyanCavanaugh** [comment]
Copilot: Let's try to fix this by specifically changing the comparability relation to revert `this` types back to the class which they are constrained to
